<a href="https://colab.research.google.com/github/elgomez7/MPM_200_Gomez/blob/main/Copy_of_assignment_9_database_management.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Packages and Scientific Computing Libraries

"""
Python Packages Overview:
-------------------------
Packages in Python are collections of modules that provide reusable code and functionality.
They allow developers to organize and share complex code libraries that can be easily imported
and used across different projects.

Key Characteristics of Packages:
- Organized hierarchically
- Contain multiple related modules
- Can be installed using package managers like pip
- Extend Python's core functionality

Three Essential Data Science Libraries:
--------------------------------------
1. Pandas (pd):
   - Primary library for data manipulation and analysis
   - Provides DataFrame: a 2D labeled data structure with columns of potentially different types
   - Key Features:
     * Reading/writing data from various sources (CSV, Excel, SQL databases)
     * Data cleaning and preprocessing
     * Complex data transformations
     * Powerful data analysis and aggregation tools

2. NumPy (np):
   - Fundamental package for scientific computing in Python
   - Introduces the ndarray (n-dimensional array) for efficient numerical operations
   - Key Features:
     * High-performance multidimensional array object
     * Mathematical and statistical functions
     * Linear algebra operations
     * Foundation for scientific and numerical computing
     * Enables fast computations on large datasets

3. Matplotlib (plt):
   - Comprehensive library for creating static, animated, and interactive visualizations
   - Provides MATLAB-like plotting interface
   - Key Features:
     * Line plots, scatter plots, bar charts, histograms
     * Customizable plot styles and layouts
     * Support for multiple plot types and graph formats
     * Integration with NumPy for data visualization

Practical Significance:
----------------------
These libraries form the backbone of data science and scientific computing in Python,
enabling complex data analysis, numerical computations, and insightful visualizations
with minimal code.
"""

In [6]:
pip install pandas

In [7]:
pip install numpy

In [8]:
pip install matplotlib

In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [10]:
df = pd.read_excel("/content/Flavivirus host associations.xlsx")

In [11]:
df.head()

sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-time fluorogenic reverse transcription po...   

                     Journal  Issue Volume   Pages  
0  Southwestern Entomologist    NaN    NaN     NaN  
1                        NaN    NaN    NaN     NaN  
2     Zoonoses Public Health   56.0    6-7  338-56  
3     Zoonoses Public Health   56.0    6-7  338-56  
4         J Vet Diagn Invest   24.0      5  959-63

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  570 non-null    int64  
 1   Virus                   570 non-null    object 
 2   Scientific Name         570 non-null    object 
 3   Authors                 570 non-null    object 
 4   Year                    430 non-null    object 
 5   Title/accession number  570 non-null    object 
 6   Journal                 425 non-null    object 
 7   Issue                   415 non-null    float64
 8   Volume                  404 non-null    object 
 9   Pages                   404 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 44.7+ KB


In [13]:
df['Virus'].nunique()

30

In [14]:
df['Virus'].unique()

array(['Apoi virus', 'Bagaza virus', 'Banzi virus', 'Bouboui virus',
       'Bussuquara virus', 'Cacipacore virus', 'Dakar bat virus',
       'Dengue virus', 'Entebbe bat virus', 'Iguape virus',
       'Ilheus virus', 'Japanese encephalitis virus', 'Koutango virus',
       'Kunjin virus', 'Kyasanur forest virus', 'Louping-ill virus',
       'Modoc virus', 'Murray Valley encephalitis virus',
       'Omsk hemorrhagic fever virus', 'Rio Bravo virus', 'Rocio virus',
       'St. Louis Encephalitis virus', 'Tembusu virus',
       'Tickborne encephalitis virus', 'Uganda S virus', 'Usutu virus',
       'Wesselsbron virus', 'West Nile virus', 'Yellow fever virus',
       'Zika virus'], dtype=object)

In [15]:
df['Virus'].value_counts()

Virus
West Nile virus                     194
Tickborne encephalitis virus         84
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yellow fever virus                   11
Rio Bravo virus                       9
Zika virus                            7
Uganda S virus                        7
Entebbe bat virus                     6
Kyasanur forest virus                 6
Dakar bat virus                       5
Ilheus virus                          5
Apoi virus                            4
Modoc virus                           3
Murray Valley encephalitis virus      3
Wesselsbron virus                     3
Omsk hemorrhagic fever virus          3
Kunjin virus                          3
Bussuquara virus                      3
Bagaza virus                          2
Banzi virus                           2
Rocio virus                           2
Tembusu virus                         2
Koutango virus                        1
Iguape virus                          1
Bouboui virus                         1
Cacipacore virus                      1
Name: count, dtype: int64

In [16]:
df['Scientific Name'].nunique()

419

## Using pandas functions that you have seen in the class address following questions
1. How many entries are there in the dataset?
2. How many viruses are present in the dataset?
3. List those viruses.
4. How many rows of data are present for each virus?
5. How many host species are present in the data?
6. Make a list of host species present.
7. Count the number of unique host species for each virus.
8. Which virus has the most number of hosts?
9. Which host has most number of viruses?

1.- 570 total entries


2.- There are 30 different virus in the dataset

3.- Apoi virus, Bagaza virus, Banzi virus, Bouboui virus Bussuquara virus, Cacipacore virus, Dakar bat virus, Dengue virus, Entebbe bat virus, Iguape virus, Ilheus virus, Japanese encephalitis virus, Koutango virus, Kunjin virus, Kyasanur forest virus, Louping-ill virus, Modoc virus, Murray Valley encephalitis virus, Omsk hemorrhagic fever virus, Rio Bravo virus, Rocio virus, St. Louis Encephalitis virus, Tembusu virus, Tickborne encephalitis virus, Uganda S virus, Usutu virus, Wesselsbron virus, West Nile virus, Yellow fever virus, Zika virus.

4 Answer.-
Apoi virus - 4.
Bagaza virus - 2.
Banzi virus - 2.
Bouboui virus - 1.
Bussuquara virus - 3.
Cacipacore virus - 1.
Dakar bat virus - 5.
Dengue virus - 27.
Entebbe bat virus - 6.
Iguape virus - 1.
Ilheus virus - 5.
Japanese encephalitis virus - 25.
Koutango virus - 1.
Kunjin virus - 3.
Kyasanur forest virus - 6.
Louping-ill virus - 13.
Modoc virus - 3.
Murray Valley encephalitis virus - 3.
Omsk hemorrhagic fever virus - 3.
Rio Bravo virus - 9.
Rocio virus - 2.
St. Louis Encephalitis virus - 72.
Tembusu virus - 2.
Tickborne encephalitis virus - 84.
Uganda S virus - 7.
Usutu virus - 65.
Wesselsbron virus - 3.
West Nile virus - 194.
Yellow fever virus - 11.
Zika virus - 7.

5.- 419 host species in the data

In [17]:
#Answer for question number 6
df['Scientific Name'].unique()

array(['Rattus rattus', 'Myodes rufocanus', 'Apodemus argenteus',
       'Apodemus speciosus', 'Alectoris rufa', 'Phasianus colchicus',
       'Mastomys natalensis', 'Mesocricetus auratus', 'Papio papio',
       'Equus caballus', 'Alouatta belzebul', 'Proechimys guyannensis',
       'Chaerephon pumilus', 'Taphozous perforatus', 'Mops condylurus',
       'Scotophilus nigrita', 'Otomops sp', 'Carollia brevicauda',
       'Molossus molossus', 'Myotis nigricans', 'Natalus stramineus',
       'Rousettus leschenaultii', 'Artibeus planirostris',
       'Caluromys philander', 'Carollia perspicillata',
       'Didelphis marsupialis', 'Hylaeamys megacephalus',
       'Marmosa demerarae', 'Marmosa murina', 'Marmosops parvidens',
       'Mesomys hispidus', 'Oecomys sp', 'Philander opossum',
       'Proechimys cuvieri', 'Zygodontomys brevicauda',
       'Artibeus jamaicensis', 'Artibeus lituratus',
       'Glossophaga soricina', 'Artibeus sp', 'Uroderma sp',
       'Macaca nemestrina', 'Mus musculu

In [18]:
df['Scientific Name'].value_counts()

Scientific Name
Equus caballus             13
Mus musculus                5
Equus ferus                 5
Rattus rattus               5
Columba livia               4
                           ..
Pica nuttalli               1
Tachycineta bicolor         1
Thryomanes bewickii         1
Tyrannus verticalis         1
Zonotrichia atricapilla     1
Name: count, Length: 419, dtype: int64

In [19]:
#Answer for question 7
df['Virus'].value_counts()

Virus
West Nile virus                     194
Tickborne encephalitis virus         84
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yellow fever virus                   11
Rio Bravo virus                       9
Zika virus                            7
Uganda S virus                        7
Entebbe bat virus                     6
Kyasanur forest virus                 6
Dakar bat virus                       5
Ilheus virus                          5
Apoi virus                            4
Modoc virus                           3
Murray Valley encephalitis virus      3
Wesselsbron virus                     3
Omsk hemorrhagic fever virus          3
Kunjin virus                          3
Bussuquara virus                      3
Bagaza virus                          2
Banzi virus                           2
Rocio virus                           2
Tembusu virus                         2
Koutango virus                        1
Iguape virus                          1
Bouboui virus                         1
Cacipacore virus                      1
Name: count, dtype: int64

8.- West Nile virus is the virus with the most number of host infected.

9.-The host with the most number of viruses is Equus Caballus (Horses)

In [20]:
# Exploring Virus and Species Diversity in the Dataset

# Count total number of entries in the dataset

# Count unique viruses

# List of unique viruses

# Count of each unique virus

# Count unique host species

# List of unique host species

# Count of host species for each virus

# Most diverse viruses (viruses with most unique host species)


# Metadata

## Given that we are interested in the metadata associated with these host species, we will import datasets associated with the metadata for both bird and mammalian species.

1. Birds

In [21]:
birds_meta = pd.read_excel("/content/Handbook of the Birds of the World and BirdLife International Digital Checklist of the Birds of the World_Version_9.xlsx")
birds_meta.head()

Seq.  Subsp. Seq.             Order    Family name     Family Subfamily  \
0   1.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
1   1.0          1.0               NaN            NaN        NaN       NaN   
2   1.0          2.0               NaN            NaN        NaN       NaN   
3   1.0          3.0               NaN            NaN        NaN       NaN   
4   1.0          4.0               NaN            NaN        NaN       NaN   

  Tribe     Common name             Scientific name       Authority  \
0   NaN  Common Ostrich            Struthio camelus  Linnaeus, 1758   
1   NaN             NaN   Struthio camelus syriacus             NaN   
2   NaN             NaN    Struthio camelus camelus             NaN   
3   NaN             NaN  Struthio camelus massaicus             NaN   
4   NaN             NaN  Struthio camelus australis             NaN   

  2024 IUCN Red List category Synonyms Alternative common names  \
0                          LC      NaN                      NaN   
1                         NaN      NaN                      NaN   
2                         NaN      NaN                      NaN   
3                         NaN      NaN                      NaN   
4                         NaN      NaN                      NaN   

                                 Taxonomic source(s)    SISRecID  \
0  del Hoyo, J., Collar, N.J., Christie, D.A., El...  45020636.0   
1                                                NaN         NaN   
2                                                NaN         NaN   
3                                                NaN         NaN   
4                                                NaN         NaN   

   SpcRecID\n(used before SISRecID)    SubsppID  
0                         1016860.0         NaN  
1                               NaN  45020636_1  
2                               NaN  45020636_2  
3                               NaN  45020636_3  
4                               NaN  45020636_4

In [22]:
birds_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33265 entries, 0 to 33264
Data columns (total 17 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Seq.                             33252 non-null  float64
 1   Subsp. Seq.                      31224 non-null  float64
 2   Order                            13233 non-null  object 
 3   Family name                      13232 non-null  object 
 4   Family                           13232 non-null  object 
 5   Subfamily                        8123 non-null   object 
 6   Tribe                            3583 non-null   object 
 7   Common name                      11195 non-null  object 
 8   Scientific name                  33252 non-null  object 
 9   Authority                        13232 non-null  object 
 10  2024 IUCN Red List category      13232 non-null  object 
 11  Synonyms                         2598 non-null   object 
 12  Alternative common

The data for birds is at subspecies level. It means that there is species name represented by column `Seq.`. All rows in same number represent one species.

2. Mammals

## Similarly, read file for mammals metadata.

In [23]:
mammals_meta = pd.read_csv("/content/IUCN Mammals, Birds, Reptiles, and Amphibians.csv",encoding='latin-1')
mammals_meta.head()

Species ID   Kingdom    Phylum     Class             Order       Family  \
0     42641.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
1  22687170.0  ANIMALIA  CHORDATA      AVES  CAPRIMULGIFORMES  TROCHILIDAE   
2     17879.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
3  47760825.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA   CRICETIDAE   
4     42656.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA  ABROCOMIDAE   

          Genus    Species                  Authority  Infraspecific rank  \
0     Abditomys   latidens            (Sanborn, 1952)                 NaN   
1      Abeillia   abeillei  (Lesson & DeLattre, 1839)                 NaN   
2   Abeomelomys      sevia    (Tate & Archbold, 1935)                 NaN   
3  Abrawayaomys    ruschii         Cunha & Cruz, 1979                 NaN   
4      Abrocoma  bennettii           Waterhouse, 1837                 NaN   

   ...                             Synonyms  \
0  ...                                  NaN   
1  ...                                  NaN   
2  ...  Melomys sevia|Pogonomelomys sevia     
3  ...                                  NaN   
4  ...                                  NaN   

                                  Common names (Eng)  Common names (Fre)  \
0                            Luzon Broad-toothed Rat                 NaN   
1                        Emerald-chinned Hummingbird                 NaN   
2  Menzies' Mouse, Menzies's Mouse, Highland Brus...                 NaN   
3                                       Ruschi's Rat                 NaN   
4                           Bennett's Chinchilla Rat                 NaN   

  Common names (Spa) Red List status Red List criteria  \
0                NaN              DD               NaN   
1                NaN              LC               NaN   
2                NaN              LC               NaN   
3                NaN              LC               NaN   
4                NaN              LC               NaN   

  Red List criteria version Year assessed Population trend  Petitioned  
0                       3.1        2008.0          unknown           N  
1                       3.1        2012.0       decreasing           N  
2                       3.1        2016.0          unknown           N  
3                       3.1        2016.0          unknown           N  
4                       3.1        2016.0          unknown           N  

[5 rows x 23 columns]

In [24]:
mammals_meta['Scientific name'] = mammals_meta['Genus'] + ' ' + mammals_meta['Species']

In [25]:
mammals_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16081 entries, 0 to 16080
Data columns (total 24 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Species ID                 15961 non-null  float64
 1   Kingdom                    16081 non-null  object 
 2   Phylum                     16081 non-null  object 
 3   Class                      16081 non-null  object 
 4   Order                      16081 non-null  object 
 5   Family                     16081 non-null  object 
 6   Genus                      16081 non-null  object 
 7   Species                    16077 non-null  object 
 8   Authority                  15955 non-null  object 
 9   Infraspecific rank         0 non-null      float64
 10  Infraspecific name         0 non-null      float64
 11  Infraspecific authority    0 non-null      float64
 12  Stock/subpopulation        0 non-null      float64
 13  Synonyms                   3168 non-null   obj

## For both birds and mammals answer the following questions.
1. Size of the datasets
2. For birds, filter the dataset using species-level rows only. Hint. Column `Subsp. Seq.` value of 0 indicates the main species name
3. How many bird species and mammal species are present in data frames?

1.- 33265 entries for the birds dataset and 16081 entries for the mammals

In [26]:
#2
birds_meta = birds_meta[birds_meta['Subsp. Seq.'] == 0]

3.- Now the bird data has 11204 entries, and mammals has still 16081 entries.

In [27]:
birds_meta.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11204 entries, 0 to 32808
Data columns (total 17 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Seq.                             11204 non-null  float64
 1   Subsp. Seq.                      11204 non-null  float64
 2   Order                            11204 non-null  object 
 3   Family name                      11204 non-null  object 
 4   Family                           11204 non-null  object 
 5   Subfamily                        6962 non-null   object 
 6   Tribe                            3120 non-null   object 
 7   Common name                      11195 non-null  object 
 8   Scientific name                  11204 non-null  object 
 9   Authority                        11204 non-null  object 
 10  2024 IUCN Red List category      11204 non-null  object 
 11  Synonyms                         2543 non-null   object 
 12  Alternative common name

In [28]:
mammals_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16081 entries, 0 to 16080
Data columns (total 24 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Species ID                 15961 non-null  float64
 1   Kingdom                    16081 non-null  object 
 2   Phylum                     16081 non-null  object 
 3   Class                      16081 non-null  object 
 4   Order                      16081 non-null  object 
 5   Family                     16081 non-null  object 
 6   Genus                      16081 non-null  object 
 7   Species                    16077 non-null  object 
 8   Authority                  15955 non-null  object 
 9   Infraspecific rank         0 non-null      float64
 10  Infraspecific name         0 non-null      float64
 11  Infraspecific authority    0 non-null      float64
 12  Stock/subpopulation        0 non-null      float64
 13  Synonyms                   3168 non-null   obj

# Database Integration Exercise
The main objective of this exercise is to create a comprehensive database by combining three data frames. Follow these steps  n your analysis.

## Create a concept map:  
* Illustrate the relationships between the three data frames.  
* Identify and highlight the Primary Keys in each data frame.  
* Use a diagramming tool (e.g., PowerPoint, LucidChart, or Draw.io) to create a visual representation of these connections.  
* Pay special attention to how you plan to merge the Mammals and Birds data frames.  

## Perform a vertical join of Mammals and Birds data frames:  
* Identify and select only the common columns present in both data frames.  
* Use `pd.concat()` or `df.append()` to create a larger dataframe called **metadata**.  
* Ensure the resulting dataframe has the correct number of rows and columns.  

In [29]:
bird_cols = set(birds_meta.columns)
mammal_cols = set(mammals_meta.columns)
common_cols = bird_cols.intersection(mammal_cols)

In [30]:
common_cols

{'Authority', 'Family', 'Order', 'Scientific name', 'Synonyms'}

In [31]:
cols_list = list(common_cols)

In [32]:
metadata = pd.concat([birds_meta[cols_list], mammals_meta[cols_list]], axis=0)

In [33]:
metadata.head()

Synonyms             Order     Family         Scientific name  \
0       NaN  STRUTHIONIFORMES  Ostriches        Struthio camelus   
5       NaN  STRUTHIONIFORMES  Ostriches  Struthio molybdophanes   
6       NaN  STRUTHIONIFORMES      Rheas          Rhea americana   
12      NaN  STRUTHIONIFORMES      Rheas            Rhea pennata   
16      NaN  STRUTHIONIFORMES   Tinamous      Nothocercus julius   

            Authority  
0      Linnaeus, 1758  
5     Reichenow, 1883  
6    (Linnaeus, 1758)  
12    d'Orbigny, 1834  
16  (Bonaparte, 1854)

In [34]:
metadata = metadata.rename(columns={'Scientific name': 'Scientific Name'})

## Merge the metadata dataframe with the Flavivirus dataframe:  
* Experiment with different join types: `inner`, `outer`, `left`, and `right`.  
* For each join type, record the resulting dataframe's shape (rows and columns).  
* Determine the **theoretically correct size of the merged dataframe if done manually**.  
* Compare the results of each join type with the expected outcome.  
* Choose the most appropriate join type based on your analysis.  

In [35]:
merged_result = pd.merge(df, metadata, on='Scientific Name', how='inner')

In [36]:
merged_result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 872 entries, 0 to 871
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  872 non-null    int64  
 1   Virus                   872 non-null    object 
 2   Scientific Name         872 non-null    object 
 3   Authors                 872 non-null    object 
 4   Year                    658 non-null    object 
 5   Title/accession number  872 non-null    object 
 6   Journal                 654 non-null    object 
 7   Issue                   644 non-null    float64
 8   Volume                  633 non-null    object 
 9   Pages                   633 non-null    object 
 10  Synonyms                181 non-null    object 
 11  Order                   872 non-null    object 
 12  Family                  872 non-null    object 
 13  Authority               820 non-null    object 
dtypes: float64(1), int64(1), object(12)
memory

In [37]:
merged_result1 = pd.merge(df, metadata, on='Scientific Name', how='outer')

In [38]:
merged_result1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27514 entries, 0 to 27513
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  881 non-null    float64
 1   Virus                   881 non-null    object 
 2   Scientific Name         27510 non-null  object 
 3   Authors                 881 non-null    object 
 4   Year                    662 non-null    object 
 5   Title/accession number  881 non-null    object 
 6   Journal                 657 non-null    object 
 7   Issue                   647 non-null    float64
 8   Volume                  636 non-null    object 
 9   Pages                   636 non-null    object 
 10  Synonyms                5768 non-null   object 
 11  Order                   27505 non-null  object 
 12  Family                  27505 non-null  object 
 13  Authority               27357 non-null  object 
dtypes: float64(2), object(12)
memory usage

In [39]:
merged_result2 = pd.merge(df, metadata, on='Scientific Name', how='left')

In [40]:
merged_result2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 881 entries, 0 to 880
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  881 non-null    int64  
 1   Virus                   881 non-null    object 
 2   Scientific Name         881 non-null    object 
 3   Authors                 881 non-null    object 
 4   Year                    662 non-null    object 
 5   Title/accession number  881 non-null    object 
 6   Journal                 657 non-null    object 
 7   Issue                   647 non-null    float64
 8   Volume                  636 non-null    object 
 9   Pages                   636 non-null    object 
 10  Synonyms                181 non-null    object 
 11  Order                   872 non-null    object 
 12  Family                  872 non-null    object 
 13  Authority               820 non-null    object 
dtypes: float64(1), int64(1), object(12)
memory

In [41]:
merged_result3 = pd.merge(df, metadata, on='Scientific Name', how='right')

In [42]:
merged_result3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27505 entries, 0 to 27504
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  872 non-null    float64
 1   Virus                   872 non-null    object 
 2   Scientific Name         27501 non-null  object 
 3   Authors                 872 non-null    object 
 4   Year                    658 non-null    object 
 5   Title/accession number  872 non-null    object 
 6   Journal                 654 non-null    object 
 7   Issue                   644 non-null    float64
 8   Volume                  633 non-null    object 
 9   Pages                   633 non-null    object 
 10  Synonyms                5768 non-null   object 
 11  Order                   27505 non-null  object 
 12  Family                  27505 non-null  object 
 13  Authority               27357 non-null  object 
dtypes: float64(2), object(12)
memory usage

### Create a pivote table,
1. Rows are virueses  
2. Columns are Taxonomic orders  
3. Values are number of unique host species  

In [47]:
pivot = merged_result2.pivot_table(index='Virus', columns='Order', values='Scientific Name', aggfunc='nunique')

In [48]:
print (pivot)

Order                             ACCIPITRIFORMES  ANSERIFORMES  \
Virus                                                             
Apoi virus                                    NaN           NaN   
Bagaza virus                                  NaN           NaN   
Banzi virus                                   NaN           NaN   
Bouboui virus                                 NaN           NaN   
Bussuquara virus                              NaN           NaN   
Cacipacore virus                              NaN           NaN   
Dakar bat virus                               NaN           NaN   
Dengue virus                                  NaN           NaN   
Entebbe bat virus                             NaN           NaN   
Iguape virus                                  NaN           NaN   
Ilheus virus                                  NaN           NaN   
Japanese encephalitis virus                   NaN           NaN   
Koutango virus                                NaN           Na